kyo-core/shared/src/main/scala/kyo/aborts.scala

```diff
@@ -4,95 +4,90 @@ import kyo.core.*
 
 object Aborts:
-    private case object aborts extends Aborts[Any]
-    def apply[V]: Aborts[V] = aborts.asInstanceOf[Aborts[V]]
 
-    def fail[V](value: V)(using t: Tag[Aborts[V]]): Nothing < Aborts[V] =
-        Aborts[V].fail(value)
+    import internal.*
 
-    def when[V](b: Boolean)(value: => V)(using Tag[Aborts[V]]): Unit < Aborts[V] =
+    def fail[V](v: V): Nothing < Aborts[V] =
+        DoAbort.suspend(v).asInstanceOf[Nothing < Aborts[V]]
+
+    def when[V](b: Boolean)(value: => V): Unit < Aborts[V] =
         if b then fail(value)
         else ()
 
-    def get[V, T](e: Either[V, T])(using Tag[Aborts[V]]): T < Aborts[V] =
-        Aborts[V].get(e)
-
-    extension [V](self: Aborts[V])
-
-        def fail[T <: V](value: T)(using t: Tag[Aborts[T]]): Nothing < Aborts[T] =
-            self.suspend[Nothing](Left(value))
+    def get[V, T](e: Either[V, T]): T < Aborts[V] =
+        e match
+            case Right(v) => v
+            case Left(v)  => DoAbort.suspend(v).asInstanceOf[Nothing < Aborts[V]]
 
-        def when(b: Boolean)(value: => V)(using Tag[Aborts[V]]): Unit < Aborts[V] =
-            if b then fail(value)
-            else ()
+    class RunDsl[V]():
+        def apply[T: Flat, S, VS, VR](v: T < (Aborts[VS] & S))(
+            using
+            h: HasAborts[V, VS] { type Remainder = VR },
+            ct: ClassTag[V]
+        ): Either[V, T] < (VR & S) =
+            DoAbort.handle(handler)(ct, v).asInstanceOf[Either[V, T] < (VR & S)]
+    end RunDsl
 
-        def get[T](e: Either[V, T])(using Tag[Aborts[V]]): T < Aborts[V] =
-            e match
-                case Right(value) => value
-                case Left(e)      => fail(e)
+    def run[V]: RunDsl[V] = RunDsl()
 
-        def catching[T, S](v: => T < S)(
-            using
-            ClassTag[V],
-            Tag[Aborts[V]]
+    class CatchingDsl[V]():
+        def apply[T: Flat, S](v: => T < S)(
+            using ct: ClassTag[V]
         ): T < (Aborts[V] & S) =
-            IOs.handle(v) {
-                case ex: V =>
-                    fail(ex.asInstanceOf[V])
+            IOs.catching(v) {
+                case ex: V => Aborts.fail(ex)
             }
+    end CatchingDsl
 
-        def run[T: Flat, S, VS, VR](v: T < (Aborts[VS] & S))(
-            using
-            HasAborts[V, VS] { type Remainder = VR },
-            Tag[Aborts[V]],
-            ClassTag[V]
-        ): Either[V, T] < (S & VR) =
-            self.handle(handler)((), v).asInstanceOf[Either[V, T] < (S & VR)]
+    def catching[V]: CatchingDsl[V] = CatchingDsl()
 
-        private def handler(using ClassTag[V], Tag[Aborts[V]]) =
-            new ResultHandler[Unit, self.Command, Aborts[V], [T] =>> Either[V, T], Any]:
-                def done[T](st: Unit, v: T) = Right(v)
+    private object internal:
 
-                override inline def accepts[T, U](self: Tag[T], other: Tag[U]): Boolean =
-                    self == other || self.parse <:< other.parse
+        val handler =
```

Comment:
This change also ended up saving the handler allocation! The `ClassTag` is passed as the state of the handler.